src/rust/src/tracing_ffi.rs

```diff
@@ -0,0 +1,552 @@
+}
+
+impl<L, S> ReloadHandle for Handle<L, S>
+where
+    L: From<EnvFilter> + Layer<S> + 'static,
+    S: Subscriber,
+{
+    fn reload(&self, new_filter: EnvFilter) -> Result<(), reload::Error> {
+        self.reload(new_filter)
+    }
+}
+
+pub struct TracingHandle {
+    _file_guard: Option<WorkerGuard>,
+    reload_handle: Box<dyn ReloadHandle>,
+}
+
+#[no_mangle]
+pub extern "C" fn tracing_init(
+    #[cfg(not(target_os = "windows"))] log_path: *const u8,
+    #[cfg(target_os = "windows")] log_path: *const u16,
+    log_path_len: usize,
+    initial_filter: *const c_char,
+    log_timestamps: bool,
+) -> *mut TracingHandle {
+    let initial_filter = unsafe { CStr::from_ptr(initial_filter) }
+        .to_str()
+        .expect("initial filter should be a valid string");
+
+    if log_path.is_null() {
+        tracing_init_stdout(initial_filter, log_timestamps)
+    } else {
+        let log_path = unsafe { slice::from_raw_parts(log_path, log_path_len) };
+
+        #[cfg(not(target_os = "windows"))]
+        let log_path = Path::new(OsStr::from_bytes(log_path));
+
+        #[cfg(target_os = "windows")]
+        let log_path = Path::new(OsString::from_wide(log_path));
+
+        tracing_init_file(log_path, initial_filter, log_timestamps)
+    }
+}
+
+fn tracing_init_stdout(initial_filter: &str, log_timestamps: bool) -> *mut TracingHandle {
+    let builder = tracing_subscriber::fmt()
+        .with_ansi(true)
+        .with_env_filter(initial_filter);
+
+    let reload_handle = if log_timestamps {
+        let builder = builder.with_filter_reloading();
+        let reload_handle = builder.reload_handle();
+        builder.init();
+        Box::new(reload_handle) as Box<dyn ReloadHandle>
+    } else {
+        let builder = builder.without_time().with_filter_reloading();
+        let reload_handle = builder.reload_handle();
+        builder.init();
+        Box::new(reload_handle) as Box<dyn ReloadHandle>
+    };
+
+    Box::into_raw(Box::new(TracingHandle {
+        _file_guard: None,
+        reload_handle,
+    }))
+}
+
+fn tracing_init_file(
+    log_path: &Path,
+    initial_filter: &str,
+    log_timestamps: bool,
+) -> *mut TracingHandle {
+    let file_appender =
+        tracing_appender::rolling::never(log_path.parent().unwrap(), log_path.file_name().unwrap());
+    let (non_blocking, file_guard) = tracing_appender::non_blocking(file_appender);
+
+    let builder = tracing_subscriber::fmt()
+        .with_ansi(false)
+        .with_env_filter(initial_filter)
+        .with_writer(non_blocking);
+
+    let reload_handle = if log_timestamps {
+        let builder = builder.with_filter_reloading();
+        let reload_handle = builder.reload_handle();
+        builder.init();
+        Box::new(reload_handle) as Box<dyn ReloadHandle>
+    } else {
+        let builder = builder.without_time().with_filter_reloading();
+        let reload_handle = builder.reload_handle();
+        builder.init();
```

Comment:
`Builder::init` can panic, potentially causing UB if the panic unwinds across an FFI boundary.